ProcessLib/RichardsMechanics/RichardsMechanicsFEM-impl.h

```diff
@@ -0,0 +1,832 @@
+
+        auto C = _ip_data[ip].updateConstitutiveRelation(t, x_position, dt, u,
+                                                         temperature);
+
+        local_Jac
+            .template block<displacement_size, displacement_size>(
+                displacement_index, displacement_index)
+            .noalias() += B.transpose() * C * B * w;
+
+        double const rho = rho_SR * (1 - porosity) + porosity * rho_LR;
+        local_rhs.template segment<displacement_size>(displacement_index)
+            .noalias() -=
+            (B.transpose() * sigma_eff - N_u_op.transpose() * rho * b) * w;
+
+        //
+        // displacement equation, pressure part
+        //
+        Kup.noalias() += B.transpose() * alpha * S_L * identity2 * N_p * w;
+
+        /* For future implementation including swelling.
+        double const dsigma_eff_dp_cap = -K_intrinsic * m_swell * n *
+                                         std::pow(S_L, n - 1) * dS_L_dp_cap *
+                                         identity2;
+        local_Jac
+            .template block<displacement_size, pressure_size>(
+                displacement_index, pressure_index)
+            .noalias() -= B.transpose() * dsigma_eff_dp_cap * N_p * w;
+        */
+
+        local_Jac
+            .template block<displacement_size, pressure_size>(
+                displacement_index, pressure_index)
+            .noalias() -= B.transpose() * alpha *
+                          (S_L + p_cap_ip * dS_L_dp_cap) * identity2 * N_p * w;
+
+        local_Jac
+            .template block<displacement_size, pressure_size>(
+                displacement_index, pressure_index)
+            .noalias() +=
+            N_u_op.transpose() * porosity * rho_LR * dS_L_dp_cap * b * N_p * w;
+        //
+        // pressure equation, displacement part.
+        //
+        Kpu.noalias() += N_p.transpose() * S_L * rho_LR * alpha *
+                         identity2.transpose() * B * w;
+
+        //
+        // pressure equation, pressure part.
+        //
+        laplace_p.noalias() +=
+            dNdx_p.transpose() * rho_LR * k_rel * K_over_mu * dNdx_p * w;
+
+        double const a0 = (alpha - porosity) / K_SR;
+        double const specific_storage =
+            dS_L_dp_cap * (p_cap_ip * S_L * a0 - porosity) +
+            S_L * (porosity / K_LR + S_L * a0);
+
+        double const dspecific_storage_dp_cap =
+            d2S_L_dp_cap_2 * (p_cap_ip * S_L * a0 - porosity) +
+            dS_L_dp_cap *
+                (porosity / K_LR + a0 * 3 * S_L + dS_L_dp_cap * p_cap_ip * a0);
+
+        storage_p.noalias() +=
+            N_p.transpose() * rho_LR * specific_storage * N_p * w;
+
+        local_Jac
+            .template block<pressure_size, pressure_size>(pressure_index,
+                                                          pressure_index)
+            .noalias() -= N_p.transpose() * rho_LR * p_cap_dot_ip *
+                          dspecific_storage_dp_cap * N_p * w;
+
```

Comment:
`-=` should be `+=`, because
d (alpha dot p)/dp = d (alpha)/dp dot p + alpha d(dot p)/dp.  In which  
```d (alpha)/dp dot p = d (alpha)/dp_c dp_c/dp dot p =  d (alpha)/dp_c (-1) (-dot p_c) =  d (alpha)/dp_c dot p_c)```

Reply:
What you say, is correct, though the convergence is worse. I'm looking into this, but if there is no improvement, I'll omit the term completely for now...